src/axom/spin/ImplicitGrid.hpp

```diff
@@ -215,33 +246,77 @@ class ImplicitGrid
+   * \param [in] bboxes an array of bounding boxes for each element
+   * \param [in] startIdx the first index of the first bounding box in bboxes
+   */
+  void insert(IndexType nelems,
+              const SpatialBoundingBox* bboxes,
+              IndexType startIdx = 0)
   {
     SLIC_ASSERT(m_initialized);
+    const double expansionFactor = m_expansionFactor;
+    LatticeType lattice = m_lattice;
 
-    // Note: We slightly inflate the bbox of the objects.
-    //       This effectively ensures that objects on grid boundaries are added
-    //       in all nearby grid cells.
+    BitsetType* binData[NDIMS];
+    IndexType* minBlkBins[NDIMS];
+    IndexType* maxBlkBins[NDIMS];
+    IndexType highestBins[NDIMS];
+    for(int i = 0; i < NDIMS; i++)
+    {
+      binData[i] = m_binData[i].data().data();
+      highestBins[i] = m_binData[i].set()->size() - 1;
+      minBlkBins[i] = m_minBlockBin[i].data();
+      maxBlkBins[i] = m_maxBlockBin[i].data();
+    }
 
-    bbox.expand(m_expansionFactor);
+    using AtomicPol = typename axom::execution_space<ExecSpace>::atomic_policy;
 
-    const GridCell lowerCell = m_lattice.gridCell(bbox.getMin());
-    const GridCell upperCell = m_lattice.gridCell(bbox.getMax());
+    for_all<ExecSpace>(
+      nelems,
+      AXOM_LAMBDA(axom::IndexType ibox) {
+        IndexType elemIdx = startIdx + ibox;
 
-    for(int i = 0; i < NDIMS; ++i)
-    {
-      BinBitMap& binData = m_binData[i];
+        SpatialBoundingBox scaledBox = bboxes[ibox];
+        // Note: We slightly inflate the bbox of the objects.
+        //       This effectively ensures that objects on grid boundaries are added
+        //       in all nearby grid cells.
+        scaledBox.expand(expansionFactor);
 
-      const IndexType lower =
-        axom::utilities::clampLower(lowerCell[i], IndexType());
-      const IndexType upper =
-        axom::utilities::clampUpper(upperCell[i], highestBin(i));
+        const GridCell lowerCell = lattice.gridCell(scaledBox.getMin());
+        const GridCell upperCell = lattice.gridCell(scaledBox.getMax());
 
-      for(int j = lower; j <= upper; ++j)
-      {
-        binData[j].set(idx);
-      }
-    }
+        for(int idim = 0; idim < NDIMS; idim++)
+        {
+          const IndexType lower =
+            axom::utilities::clampLower(lowerCell[idim], IndexType());
+          const IndexType upper =
+            axom::utilities::clampUpper(upperCell[idim], highestBins[idim]);
+
+          const IndexType word = elemIdx / BitsetType::BitsPerWord;
+
+          for(int j = lower; j <= upper; ++j)
+          {
+            binData[idim][j].atomicSet(elemIdx);
+#ifdef AXOM_USE_RAJA
+            RAJA::atomicMin<AtomicPol>(&minBlkBins[idim][j], word);
+            RAJA::atomicMax<AtomicPol>(&maxBlkBins[idim][j], word);
+#else
+            minBlkBins[idim][j] = std::min(minBlkBins[idim][j], word);
+            maxBlkBins[idim][j] = std::max(maxBlkBins[idim][j], word);
+#endif
```

Comment:
Could this be handled within axom's ExecutionSpace / AtomicPolicy implementation?
It seems a bit strange that AtomicPolicy is defined in Axom
>     using AtomicPol = typename axom::execution_space<ExecSpace>::atomic_policy;
(defined a few lines up)

but when we use it we suddenly need to know if we have RAJA.

Reply:
I think the atomic policies are aliased directly to the corresponding RAJA policies (and to void if axom isn't built with RAJA). We could potentially wrap the RAJA atomic operations on our side, in order to handle the axom-minus-RAJA case more cleanly.